### ISSUE TITLE
Inconsistent error handling

### ISSUE THREAD

**Issue by banderror** [redux-saga/redux-saga#632]
## TL;DR

1. Can't catch exceptions fired by `put`.
1. Unhandled exceptions are not always being caught by a custom global handler (if specified).
1. If a custom global handler is specified, the default console logging (by the saga middleware) is not being suppressed.
1. Handling exceptions of a spawned task is not considered as handling by the saga middleware.

Related issues:
- [#250 Handling uncaught exceptions](https://github.com/yelouafi/redux-saga/issues/250)
- [#374 Question: handling uncaught exceptions - spawn vs fork](https://github.com/yelouafi/redux-saga/issues/374)
- [#441 Weird (incorrect?) error handling behaviors](https://github.com/yelouafi/redux-saga/issues/441)
- [#626 Contradiction in docs whether `put` is blocking](https://github.com/yelouafi/redux-saga/issues/626)

## The problem

We have a shopping cart on our website, and the ability to store it locally and restore
afterwards at some point.

During a restore we can encounter some errors, e.g.:
- storage is not available
- restored data is broken (and can't be rendered)

We want to be able to catch and handle any of these cases 
(here and actually in other sagas in general):
- at least report the error to a backend like NewRelic
- recover from the error, don't break the whole app (don't abort the root saga)

## Initial code

Here is our code (slightly simplified, but very close to reality).

```js
function fetchCart(restaurantId) {
    // Fetches the cart and returns a Promise, which returns cart items.
}

function reportError(e) {
    // Logs errors, sends them to some backend like NewRelic, etc.
}
```

```js
const sagaMiddleware = createSagaMiddleware({
    onError: reportError
});
```

```js
function* restoreCartHandler(action) {
    const { restaurantId } = action.payload;
    try {
        const cartItems = yield call(fetchCart, restaurantId);
        yield put(cartRestored(cartItems));
    } catch (e) {
        // If something goes wrong, we clear the cart and allow user to start from the beginning.
        reportError(e);
        yield put(clearCart(restaurantId));
    }
}

function* restoreCartSaga(context) {
    yield takeEvery(RESTORE_REQUESTED, restoreCartHandler);
}
```

If the data is broken and `put` throws (e.g. from a reducer or from a component's render method),
we can't catch it in `restoreCartHandler`. `put` is async (non-blocking), and it always returns
ASAP without exceptions. Exceptions bubble up to the redux-saga middleware and
are finally console.logged as `uncaught at restoreCartHandler`.
Still, we can catch other exceptions: e.g. syntactic and from the `fetchCart` (e.g. storage
is not available).

If the data is broken, this is what I see in the browser console:

![](https://puu.sh/sj2fo/dfee968b62.png)

So here are two problems: 
1. We can't catch exceptions triggered by put and handle the case in the `restoreCartHandler`.
1. Moreover, this unhandled exception is not being caught by the global handler.

## Solution 1: wrapped `fork` and `spawn`

Based on [#250](https://github.com/yelouafi/redux-saga/issues/250) and 
[#374](https://github.com/yelouafi/redux-saga/issues/374) I tried to implement my own
`takeEvery` which could catch exceptions from `put` effects and report them.

```js
function* forkSafely(saga, ...args) {
    const task = yield fork(saga, ...args);
    task.done.catch(e => {
        console.log('Exception caught inside forkSafely', e);
    });
    return task;
}

function* spawnSafely(saga, ...args) {
    const task = yield spawn(saga, ...args);
    task.done.catch(e => {
        console.log('Exception caught inside spawnSafely', e);
    });
    return task;
}

function forkWrapped(saga, ...args) {
    return call(forkSafely, saga, ...args);
}

function spawnWrapped(saga, ...args) {
    return call(spawnSafely, saga, ...args);
}

export function* takeEverySafely(pattern, saga, ...args) {
    while (true) {
        const action = yield take(pattern);
        yield forkWrapped(saga, ...args.concat(action));
        
        // Or, instead:
        // yield spawnWrapped(saga, ...args.concat(action));
    }
}
```

```js
export function* restoreCartSaga(context) {
    yield takeEverySafely(RESTORE_REQUESTED, restoreCartHandler, context);
}
```

The result using wrapped `fork`:

![](https://puu.sh/sj44s/a4d7ad2019.png)

Neither `Exception caught inside forkSafely`, nor reporting the error from the global handler.

The result using wrapped `spawn` is the same.

So, in both cases (fork and spawn):
1. We can't catch an exception triggered by put and handle it in the handler attached to the `task.done.catch`
1. Again, this unhandled exception is not being caught by the global handler.

### Modified case: exception not in `put`

It was interesting to check what is the error handling behaviour if exceptions are fired
somewhere else in the handler, not by the `put`.

```js
function* restoreCartHandler(context, action) {
    // Let's just throw here and leave it unhandled
    throw Error('Sorry man');
}
```

```js
const sagaMiddleware = createSagaMiddleware({
    onError: e => {
        // And let's modify this one just for clarity on screen shots
        console.log('Global saga error handler', e);
    }
});
```

I was expecting that `fork` and `spawn` return a task in that case, and `task.done.catch` will work ok.

Let's see the result of using `fork`:

![](https://puu.sh/sj4Gp/934e1f79b7.png)

1. `task.done.catch` handler is not being called, no `Exception caught inside forkSafely`.
1. The exception bubbles up to the root saga, the root saga is being aborted.
1. The global handler is being called, but the default console logging is not being suppressed.

Let's see the result of using `spawn`:

![](https://puu.sh/sj57F/5329e3c1f3.png)

1. `task.done.catch` handler is being called, but after all the other handlers.
1. The exception doesn't bubble up to the root saga.
1. Despite of the fact that the exception is handled by the handler attached to the spawned task:
    - the task is still considered uncaught (`uncaught at restoreCartHandler`)
    - the global handler is still being called

There were big hopes to the "spawn-and-catch" approach, but currently it doesn't seem to be
very usable due to its implementation. "Fork-and-catch" doesn't work at all.

## Solution 2: `put().done.catch`

The idea was to check whether I can asynchronously catch and handle errors of the `put` itself,
since it seemed to behave similar to `fork` and `spawn`.

```js
function* restoreCartHandler(action) {
    const { restaurantId } = action.payload;
    try {
        const cartItems = yield call(fetchCart, restaurantId);
        const task = yield put(cartRestored(cartItems));
        task.done.catch(e => {
            console.log('Exception caught inside put', e);
        });
    } catch (e) {
        // If something goes wrong, we clear the cart and allow user to start from the beginning.
        reportError(e);
        yield put(clearCart(restaurantId));
    }
}

function* restoreCartSaga(context) {
    yield takeEvery(RESTORE_REQUESTED, restoreCartHandler);
}
```

The result is:

![](https://puu.sh/sj6Al/4fd53b5ac4.png)

1. `put` returns `undefined`, which leads to an exception inside the try block of the `restoreCartHandler`
    - this exception is being caught and reported
1. `put` is being executed as usual, with the "uncaught at..." result
    - so the original exception is not being caught, even by the global handler

## Solution 3: `put.sync`

We used `put` everywhere in the codebase as we believed that it's blocking.
Since it happens that it was not true either from the beginning, or from some version 
(we've upgraded from `0.10` to `0.12`), I decided to check `put.sync`.

```js
function* restoreCartHandler(action) {
    const { restaurantId } = action.payload;
    try {
        const cartItems = yield call(fetchCart, restaurantId);
        yield put.sync(cartRestored(cartItems));
    } catch (e) {
        // If something goes wrong, we clear the cart and allow user to start from the beginning.
        reportError(e);
        yield put(clearCart(restaurantId));
    }
}

function* restoreCartSaga(context) {
    yield takeEvery(RESTORE_REQUESTED, restoreCartHandler);
}
```

The result is finally as expected:

![](https://puu.sh/sj6WN/843236a8ce.png)

1. `put.sync` throws exception inside the try block.
1. It is being caught, reported and handled properly.
1. The exception doesn't bubble up to our global handler, and is not being logged by the middleware.

## To sum up

Seems that the only viable solution for us - to use `put.sync` for the time being.

Still, I suppose error handling could have been more consistent:
- if `put` is non-blocking, it should return a task
- exceptions fired by `put` should be able to be caught
- what's caught locally (both sync and async way), if there's no re-throw:
    - should be treated as handled
    - should not bubble up to the root
- if a custom global handler is specified, no default logging should be done

This is what basically seems right from our perspective.
Would be great to hear your opinion.

**Comment by sebinsua** [comment]
My approach to part of this problem was to avoid handling all errors with a gigantic try-catch.

See: https://github.com/sebinsua/redux-saga-helpers

I'll read through this later on...

**Comment by yelouafi** [comment]
>Solution 1: wrapped fork and spawn

Handling errors with `done.catch` wont prevent errors from bubbling. `done.catch` is meant only to be used outside of Sagas (eg monitors, root sagas). But in your case since put is already swallowing the error, no error is reported to the promise

To prevent aborted forks from killing the entire tree use `call`

```js
function* forkSafely(saga, ...args) {
   try {
     yield call(saga, ...args);
  } catch (err) {
    console.log('Exception caught inside forkSafely', e);
  }
}
```

But that wont also work with the current (weird) put behavior, because the error is swallowed (but will work with put.sync)

>put is async (non-blocking), and it always returns ASAP without exceptions.

Actually not quite true. put will report the result of the dispatch ([here](https://github.com/yelouafi/redux-saga/blob/master/src/internal/proc.js#L449)). The difference with put.sync is that the latter will wait if the result is a promise.

>if put is non-blocking, it should return a task

I agree, [the error reporting behavior](https://github.com/yelouafi/redux-saga/blob/master/src/internal/proc.js#L442-L443) of (non sync) puts  doesn't look quite right when I look at it right now (In fact very weird). It seems all correct for me to bubble the error also for normal puts (ie non sync) since we have this info before resuming the Saga. 

>what's caught locally (both sync and async way), if there's no re-throw:
>   should be treated as handled
>  should not bubble up to the root

I think that's the actual behavior (besides the weird put 'bug'). Except for fork && spawn, because I think the functions should still return a task to the caller (even aborted task)

>if a custom global handler is specified, no default logging should be done

Hmm... not sure about this. onError was introduced as a kind of post-mortem handler. That being said, I'm open to hear opinions about the desired behavior of a global error handler; for example should the global handler prevents fork errors from bubbling && killing the parent tree?

**Comment by Andarist** [comment]
> > if put is non-blocking, it should return a task
> > exceptions fired by put should be able to be caught

> I agree, [the error reporting behavior](https://github.com/yelouafi/redux-saga/blob/master/src/internal/proc.js#L442-L443) of (non sync) puts  doesn't look quite right when I look at it right now (In fact very weird). It seems all correct for me to bubble the error also for normal puts (ie non sync) since we have this info before resuming the Saga. 

Wasnt that exact behaviour of preventing error from bubbling up to the saga introduced here - https://github.com/yelouafi/redux-saga/pull/567 for mentioned there reasons?

**Comment by yelouafi** [comment]
@Andarist Yeah! but that was the other me :). 

More seriously, in #567 the idea that was 'put is non blocking so we shouldn't bubble errors'. But then, as @banderror noted we should return a task in this case, otherwise there is no way to catch errors from it. But again if we return a task, then errors will fail the entire tree just as fork/spawn do actually.

I think we shouldn't treat put as a non blocking effect (ie I think I was wrong on #567). In the actual code, put will wait (even sync) for the return value from dispatch, in fact seems like put.sync is a misleading name, because all the difference vs put alone is that the former resolve promise responses. maybe put.resolve could be a better name

**Comment by Andarist** [comment]
Returning task from `put` would be a HUGE breaking change I think. Also as you have mentioned it wouldnt help with handling both scenarios well - bubbling error and keeping the tree intact.

I agree that `put.resolve` would be a better name, I can draft a PR soon with deprecation warning on `put.sync`.

**Comment by restrry** [comment]
any action point for that?
is it official recommendation to use `put.resolve` if we want to handle such errors in saga?

**Comment by Andarist** [comment]
For now I would say yes. It for sure needs to be established before reaching 1.0, which I would like to reach eventually in a 'near' future. But for now it seems reasonable to stick to `put.resolve` here. 

Error boundaries are a quite a topic, regarding the fact that even React itself is struggling with it and potentially Fiber might be the answer. I need to dig into what they do there and whats the reasoning behind the thing they do. It could have been helpful for us here.

I think it might be more important for sagas to keep their tree intact in such a case that `put` throws under the hood.

However it potentially might lead to the inconsistent state. Do you have any example in mind when it matters? And when failing would be really a better option for us to make by default?

**Comment by skellock** [comment]
`put.resolve()` means your stack trace will be meaningful.

PUT is a sneaky little guy.  You're giving a chance for the rest of the redux environment to react (lulz).  This means anything from connected components to reducers to selectors can bring down the house.

With `put.resolve` the inner stack trace shows you exactly who caused it.  With `put`, you have to wait until next tick and you go hunting for answers.

My vote (wait, are we even voting?) is for `put.resolve` to be the default `put` behaviour.  Breaking changes be damned.  I thought it always worked like that.  😃 

What I'm doing now is:

```js
import { put, take } from 'redux-saga/effects'
const dispatch = put.resolve // ignorance is bliss
```

I know the naming is probably going to cause ya'll ulcers, but dispatching is exactly how I think of the behaviour.  Synchronous dispatching.

**Comment by Andarist** [comment]
Well, most people think about it like that. I wont throw stones at you for naming it `dispatch` 😆 

However it wont be always synchronous - no matter if you use `put` or `put.resolve`. `redux-saga` must schedule things under the hood (not too much, but still...) to accomodate 'parallel tasks' environment.

> Breaking changes be damned.

Not afraid of breaking changes! Always thought they are good for the project, keeping things non-breaking forcefully might be really constraining the project to move forward in many cases. That doesnt ofc mean that major API changes should happen easily and often.

> With put.resolve the inner stack trace shows you exactly who caused it. With put, you have to wait until next tick and you go hunting for answers.

Not sure what do you mean by that, I think logged errors are pretty much the same in both cases. Also the timing is exactly the same, the difference is that `put` will swallow the error (just log it) and `put.resolve` will throw into your saga (with a possibility to crash your app if you do not have proper `try/catch`)

**Comment by skellock** [comment]
Oh wow.  You're right.  

Previously, in React Native, if I had an error in code path in a `render()` method of a connected component, and my saga did a `PUT` that would trigger a reducer to trigger a redraw, I'd get a gloriously irrelevant stack trace of `redux-saga` internals.

Now, when I check... that's not happening anymore!

Just checked on `react-redux@5.x` and `4.x`, `react-native@0.40` and `0.41.1`, and `redux@12.1` and `0.14.3`.  Every combination works.

Well then.

Myth busted.

You enjoy your night.  I'm going drinking now.

**Comment by restrry** [comment]
>Error boundaries are a quite a topic, regarding the fact that even React itself is struggling with it and potentially Fiber might be the answer.
To be honest my concerns are not about react itself (since errors during rendering should be handled on react level), but about errors in domain logic.
Correct me if I'm wrong, but Saga by default doesn't separate action types. And actions that are dispatched by user isn't the same that are dispatched by app logic. it means that communication process relays on `put` heavily. so during `put` could throw also another forked process.

**Comment by markspolakovs** [comment]
Thanks for the `put.resolve` trick. However, if there's an exception inside the saga, the reported stack trace on React Native is completely useless (http://i.imgur.com/ephY1DJ.png).

**Comment by skellock** [comment]
Which version of React Native? >= 0.40.0?

**Comment by markspolakovs** [comment]
@skellock 0.41.2

**Comment by nshoes** [comment]
+1 halp

**Comment by Andarist** [comment]
@nshoes could u specify what help would u like to get regarding this issue?

**Comment by nshoes** [comment]
@Andarist this was catching errors for me fine with on redux-saga@0.11.0.

```javascript
export function* updateCalculator(action) {
  try {
    const { value, id, key, fromMap } = action
    yield put(updateCameraAction(key, value, id))
    yield put(updateCurrentCameraAction(id))
    yield put(updatePresentationValuesAction(key, value, fromMap))
  } catch (err) {
    yield put(messageAction(action.key, err.message))
  }
}

export function* watchUpdateCalculator() {
  yield takeEvery(UPDATE_CALCULATOR, updateCalculator)
}
```

But on `0.14.3`, it is uncaught and gets plastered onto `window`. But using `put.resolve` changes it back to the previous behavior. I asked you in Gitter about why `put` was changed to non-blocking, and then you said it was always non-blocking, so I am confused on how it works honestly. A few questions (pardon me if these are obvious):

1.) Why does passing everything to `put.resolve` catch the error properly? To give you more a little more context, our app has a validator when receiving user input:

![asdf](https://cloud.githubusercontent.com/assets/4118389/23131920/b888706a-f740-11e6-8696-2549fdecf76c.gif)

We throw actual errors and then rely on the saga to catch them and halt itself, not letting any other actions get dispatched.

2.) Why is `put` non-blocking? It dispatches an action to the reducer, which is synchronous, so why doesn't it wait until the reducer has finished it's task before moving on?

**Comment by Andarist** [comment]
>  I asked you in Gitter about why put was changed to non-blocking, and then you said it was always non-blocking, so I am confused on how it works honestly.

My comment was about its general behaviour, not the error handling itself. I strongly believe that `put` was always resolved (giving back running context to your generator) once it finally managed to `dispatch` your action. By finally I mean that it could not be immediate due to internal scheduling, but for sure within the same synchronous call stack.

You are right though that silent error swallowing was added, because at the moment it felt weird that error in your reducer or component can crash your saga. However after this got into the library the questions were raised about this behaviour and we are not 100% sure which behaviour is desired and which one should we keep. We will for sure settle our minds on it before geting v1 to you, which im pushing forward the best I can due to limited time I can spend on the code now.

In addition - there are not really that many blocking effects in the library, most are non-blocking.
More about this can be found here:
- https://redux-saga.github.io/redux-saga/docs/Glossary.html
- https://redux-saga.github.io/redux-saga/docs/api/index.html#blocking--nonblocking

> 2.) Why is put non-blocking? It dispatches an action to the reducer, which is synchronous, so why doesn't it wait until the reducer has finished it's task before moving on?

It waits, it just swallows the error for now, which `put.resolve` does not do. I dont if I can explain it better than [the code](https://github.com/redux-saga/redux-saga/blob/3b90d33d27cb1a5aa99e8672b47cf8e496d686f6/src/internal/proc.js#L430-L445) itself, let me show you:
```javascript
  function runPutEffect({channel, action, resolve}, cb) {
    /**
      Schedule the put in case another saga is holding a lock.
      The put will be executed atomically. ie nested puts will execute after
      this put has terminated.
    **/
    asap(() => {
      let result
      try {
        result = (channel ? channel.put : dispatch)(action)
      } catch(error) {
        // If we have a channel or `put.resolve` was used then bubble up the error.
        if (channel || resolve) return cb(error, true)
        log('error', `uncaught at ${name}`, error.stack || error.message || error)
      }

      if(resolve && is.promise(result)) {
        resolvePromise(result, cb)
      } else {
        return cb(result)
      }
    })
    // Put effects are non cancellables
  }
```
The only non-obvious thing here might be `cb` - its just a function which propagates the result properly through the code, so it can be injected back to your generator.
`cb(result)` - means this is a proper result and it will get injected by `it.next(result)` later
`cb(error, true)` - second param being true means this is an error, so it will later inject it later with `it.throw(error)`

If you have any further questions, please ask, I'll try to explain things to the best of my capabilities :)

**Comment by nshoes** [comment]
> My comment was about its general behavior, not the error handling itself. I strongly believe that put was always resolved (giving back running context to your generator) once it finally managed to dispatch your action. By finally I mean that it could not be immediate due to internal scheduling, but for sure within the same synchronous call stack.

Alright, that makes sense. Thanks.

> You are right though that silent error swallowing was added, because at the moment it felt weird that error in your reducer or component can crash your saga. However after this got into the library the questions were raised about this behaviour and we are not 100% sure which behaviour is desired and which one should we keep. We will for sure settle our minds on it before geting v1 to you, which im pushing forward the best I can due to limited time I can spend on the code now.

I think that it felt natural that it halted the generator, and I would assume more people feel the same.

> In addition - there are not really that many blocking effects in the library, most are non-blocking.
> More about this can be found here:

I see, thanks. After thinking about it more it makes sense it's non-blocking. Obviously the reducer has to finish whatever it's working on before taking the next action. The mental model I had was that the saga followed the async/sync behavior of what it was doing; meaning a `put` would be blocking, a `call` could be blocking or non-blocking depending on what the function is actually doing.

The source for `runPutEffect` makes a lot more sense, thanks. 

I would like to stress that it felt natural that an error would halt the saga with `put`. Would love to hear some feedback from others. Thanks for the feedback @Andarist.

**Comment by abrkn** [comment]
I've been having this issue as well when converting a project to `redux-saga`. If a reducer throws an error, the program is in an unknown state and must terminate.

**Comment by Andarist** [comment]
I've decided to restore old behaviour - we shouldn't swallow errors by default. Gonna do it in upcoming v1 release. I'm really sorry that u had to wait for so long for this, but I guess I had to mature to this decision.

**Comment by kyriiherman** [comment]
@Andarist so is this description invalid for now? https://redux-saga.js.org/docs/api/index.html#putresolveaction

**Comment by restrry** [comment]
valid. now error bubbles for both `put()` and `put.resolve()`

**Comment by kyriiherman** [comment]
@restrry sorry I missclicked, in API docs for put action said:
https://redux-saga.js.org/docs/api/index.html#putaction
>>This effect is non-blocking and any errors that are thrown downstream (e.g. in a reducer) will not bubble back into the saga. 
But you said that errors bubbles for `put()`.

**Comment by gsklee** [comment]
@Andarist does that mean the docs need to be updated?

**Comment by Andarist** [comment]
They have been updated - https://github.com/redux-saga/redux-saga/pull/1922/files . Didn't have time to do a release with those changes though. I'll try to do it this week.
